File: algorithms/go/1137-n-th-tribonacci-number/n-th-tribonacci-number.go
package problem1137

func tribonacci(n int) int {
	if n == 0 {
		return 0
	}
	if n <= 2 {
		return 1
	}
	p := 0
	q := 0
	r := 1
	s := 1
	for i := 3; i <= n; i++ {
		p = q
		q = r
		r = s
		s = p + q + r
	}
	return s
}

File: algorithms/python/1137_n_th_tribonacci_number/n_th_tribonacci_number.py
"""
n th tribonacci number

第 N 个泰波那契数
"""


def tribonacci(n: int) -> int:
    if n == 0:
        return 0
    if n <= 2:
        return 1
    p, q, r, s = 0, 0, 1, 1
    for i in range(3, n + 1):
        p = q
        q = r
        r = s
        s = p + q + r
    return s
